stackrox/stackrox
# DO NOT MERGE ROX-13819: roxctl command to manage bolt database files

## Description

We currently are having an issue where the assumption is that we are dealing with invalid entries within the groups store, specifically the following ([the source of the log is from the groups datastore which tries to self-recover and delete invalid entries](https://github.com/stackrox/stackrox/blob/d4e5f6c5cfa4381d5684a155fdfc7f7dac6055b0/central/group/datastore/datastore_impl.go#L223-L250)):
```
* group /* 7 unknown bytes */
8: 1869771883
14: 121
has no ID set and cannot be deleted 
```

However, dealing with corrupt data is quite tricky. The data also seems to be carried over multiple releases, so the backup that is used is also suspected to contain the corrupted entries.

For helping investigating this issue, we've created custom roxctl commands, which aim to provide the following:
- Insight of _which_ values are stored within a specific bucket, to figure out _which proto message_ is stored within there, to better reproduce the issue at hand.
- Re-creation of the bucket, leaving out any previously corrupted data entries.

The new commands will be `roxctl central db bolt list / re-create groups` respectively, where `list` will provide the ability to list a _specific_ bucket, and `re-create groups` to re-create the groups bucket.

Under the hood, the commands will be executed on a bolt database file. The file may be either contained within a stackrox database backup ZIP, or within a volume of a deployment. Internally, we will simply use [the bolt package](https://github.com/etcd-io/bbolt) to interact with bolt.

**Note: This change is not meant to be merged to main. It is issue-specific. Since we are closing in on deprecating the rocksdb/bolt stores, the effort shouldn't be taken to maintain this for the next two releases.**

## Checklist
~~- [ ] Evaluated and added CHANGELOG entry if required~~
~~- [ ] Determined and documented upgrade steps~~
~~- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~~

If any of these don't apply, please comment below.

## Testing Performed

- see CI for unit tests.

Manual tests executed:

1. Create a backup by executing the following command:
```shell
roxctl central backup --output . --endpoint <endpoint>:443  --password <password>
```
3. Unzip the backup file. You should find the following structure:
```shell
ls                                                                                                                                                                                                                                                                                                                    
bolt.db    group    keys    migration_version.yaml    rocks.db
```
5. Use the bolt.db file to list the entries of the groups2 bucket:
```shell
roxctl central db bolt list -f bolt.db  --details                                                                                                                                                                                                                                                                   

String value for key io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5:

d
$d01d0881-81e3-468d-ad91-77f568943eb0"<io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5Admin
Proto message for key io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5:
props: <
  id: "io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5"
  auth_provider_id: "d01d0881-81e3-468d-ad91-77f568943eb0"
>
role_name: "Admin"

Hex value for key io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5:
0a640a2464303164303838312d383165332d343638642d616439312d373766353638393433656230223c696f2e737461636b726f782e617574687a2e67726f75702e35373465646230362d346137382d343032612d626564642d3464386365343033386165351a0541646d696e
```
6. (optional): In case you want to verify the hex value, you can use this code snippet to unmarshal the hex value into a proto message:
```go
	bytes, err := hex.DecodeString("0a640a2464303164303838312d383165332d343638642d616439312d373766353638393433656230223c696f2e737461636b726f782e617574687a2e67726f75702e35373465646230362d346137382d343032612d626564642d3464386365343033386165351a0541646d696e")
	assert.NoError(t, err)
	var msg storage.Group
	err = msg.Unmarshal(bytes)
	assert.NoError(t, err)
	fmt.Print(proto.MarshalTextString(&msg))
```
or the following command (requires [protoscope](https://github.com/protocolbuffers/protoscope)):
```shell
echo \
  '0a640a2464303164303838312d383165332d343638642d616439312d373766353638393433656230223c696f2e737461636b726f782e617574687a2e67726f75702e35373465646230362d346137382d343032612d626564642d3464386365343033386165351a0541646d696e' |\
  xxd -r -ps |\
  protoscope
```
7. Re-create the groups bucket by executing the following (you might also want to first run it with the `--dry-run` flag):
```shell
roxctl central db bolt re-create groups -f bolt.db (optional: --dry-run)
```
8. Since we did not have any invalid entries, you should expect no output. Listing the bucket should give you the same entries as beforehand:
```shell
roxctl central db bolt list -b groups2 -f bolt.db                                                                                                                                                                                                                                                                     

String value for key io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5:

d
$d01d0881-81e3-468d-ad91-77f568943eb0"<io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5Admin
Proto message for key io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5:
props: <
  id: "io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5"
  auth_provider_id: "d01d0881-81e3-468d-ad91-77f568943eb0"
>
role_name: "Admin"

Hex value for key io.stackrox.authz.group.574edb06-4a78-402a-bedd-4d8ce4038ae5:
0a640a2464303164303838312d383165332d343638642d616439312d373766353638393433656230223c696f2e737461636b726f782e617574687a2e67726f75702e35373465646230362d346137382d343032612d626564642d3464386365343033386165351a0541646d696e

```

From here on, you may want to re-zip the archive, and restore the database within central using the roxctl command.
